presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergFileWriterFactory.java

```diff
@@ -131,9 +132,40 @@ private IcebergFileWriter createParquetWriter(
             HdfsContext hdfsContext,
             MetricsConfig metricsConfig)
     {
+//        class Local
+//        {
+//            Types.NestedField updateAllChildren(Types.NestedField parent)
+//            {
+//                if (parent.type().isNestedType()) {
+//                    ((org.apache.iceberg.types.Type.NestedType) parent.type()).fields() {
+//
+//                    }
+//                }
+//            }
+//        }
+
         List<String> fileColumnNames = icebergSchema.columns().stream()
                 .map(Types.NestedField::name)
                 .collect(toImmutableList());
+
+        List<Types.NestedField> updatedFields = icebergSchema.columns().stream()
+                .map(field -> {
+                    if (field.type() instanceof Types.TimestampType) {
+                        return Types.NestedField.of(
+                                field.fieldId(),
+                                field.isOptional(),
+                                field.name(),
+                                Types.fromPrimitiveString(Types.TimestampType.withoutZone().toString()));
+                    }
+
+                    if (field.type().isNestedType()) {
+                        return field;
+                    }
+
+                    return field;})
+                .collect(Collectors.toList());
+        icebergSchema = new Schema(updatedFields);
```

Comment:
As far as I can see, the `icebergSchema` has no need to be translated here. Seems that the information loss of `adjustToUTC` is omitted when flushing the parquet message type to the parquet file's footer, and vice versa reading parquet message type from the parquet file's footer metadata.

Referring to `MessageTypeConverter.toParquetSchema()` and `MetadataReader.readTypeSchema()`.

Reply:
I see your point. I think this issue ultimately boils down to the Apache `OriginalType` class missing any parameter information, and the conversion methods basically just defaulting to `isAdjustedToUTC`.  The `PrimitiveType` class we use in `messageTypeConverter` has a logical type annotation, but there is no way to directly build a `LogicalType` from this. 

Within the Apache library, there exists an interface `LogicalTypeAnnotationVisitor`, and it's inheritor `LogicalTypeConverterVisitor` does exactly what we need here, however it is private to `ParquetMetadataConverter`.

This method can only be used through `toParquetMetadata`, which I'm not sure why we aren't currently making use of. Instead of setting the schema of our file metadata using a custom method that currently isn't working for logical types, we could just convert the entire `FileMetaData` object using the method that Apache already provides us.

The other solution would be to copy the implementation of 'LogicalTypeConverterVisitor' into our message type converter so we can get types from annotations.

TLDR: Apache doesn't expose a method for us to keep going with our custom converter implementation, we can either switch to using their meta data converter of copy some of their protected code.

Reply:
Yes, you are right. Our own implementation omitted the logical type annotation. Besides the methods you mentioned above which may introduce a lot of change, another shortcut way maybe that we do special handle for the primitive type with logical type annotation of `TimestampLogicalTypeAnnotation` in `TypeVisitor.visit(primitiveType)`, set logical type for it's result `SchemaElement`. And do the opposite thing in `MetadataReader.readTypeSchema(...)`. This is not a universal method, but it can be proven to be feasible.

Reply:
I tried coding it up this way on a separate branch...

```
            public void visit(PrimitiveType primitiveType)
            {
                SchemaElement element = new SchemaElement(primitiveType.getName());
                element.setRepetition_type(toParquetRepetition(primitiveType.getRepetition()));
                element.setType(getType(primitiveType.getPrimitiveTypeName()));

                if (primitiveType.getOriginalType() != null) {
                    element.setConverted_type(getConvertedType(primitiveType.getOriginalType()));
                    LogicalTypeAnnotation annotation = primitiveType.getLogicalTypeAnnotation();

                    if (annotation instanceof LogicalTypeAnnotation.TimestampLogicalTypeAnnotation) {
                        TimeUnit unit = TimeUnit.MICROS(new MicroSeconds());
                        switch(((LogicalTypeAnnotation.TimestampLogicalTypeAnnotation) annotation).getUnit()) {
                            case NANOS:
                                unit = TimeUnit.NANOS(new NanoSeconds());
                                break;
                            case MILLIS:
                                unit = TimeUnit.MILLIS(new MilliSeconds());
                                break;
                        }

                        element.setLogicalType(LogicalType.TIMESTAMP(
                                new TimestampType(((LogicalTypeAnnotation.TimestampLogicalTypeAnnotation) annotation).isAdjustedToUTC(), unit)));
                    }
                }
```

This could obviously be separated into another method if need be. 

I think this will work but wouldn't it be better to use the method Apache already provides? What if in the future we run into a similar problem with another type? We are maintaining a custom type converter that doesn't have any custom functionality apart from the provided one as far as I can tell.

Reply:
> wouldn't it be better to use the method Apache already provides?

Of course, that would be better.